src/services/application.js

```diff
@@ -18,6 +38,160 @@ class CompanyApplicationService {
         delete retVal.password;
         return retVal;
     }
+
+    findById(id) {
+        return CompanyApplication.findById(id).exec();
+    }
+
+    /**
+     *
+     * @param {*} limit - Number of documents to return
+     * @param {*} offset - where to start the query (pagination - how many documents to skip, NOT how many pages to skip)
+     *
+     * @returns {applications, docCount}
+     */
+    async findAll(limit, offset, sortingOptions) {
+
+        const docCount = await CompanyApplication.estimatedDocumentCount();
+
+        return {
+            docCount,
+            applications: await Promise.all([...(await CompanyApplication.find({})
+                .sort(sortingOptions || { submittedAt: "desc" })
+                .skip(offset)
+                .limit(limit)
+                .exec()
+            )]
+                .map(async (application) => ({
+                    ...application.toObject(),
+                    state: (await CompanyApplication.findById(application._id).exec()).state,
+                }))),
+        };
+    }
+
+    buildFiltersQuery({
+        companyName,
+        submissionDate,
+    }) {
+        const filterQueries = [];
+
+        if (companyName) {
+            filterQueries.push({
+                // This allows for partial text matching.
+                // If only full-text is needed, use $text query instead, which uses
+                // a text index and is more performant
+                companyName: new RegExp(companyName, "gi"),
+            });
+        }
+
+        if (submissionDate) {
+
+            const filter = { $and: [] };
+
+            if (submissionDate.to) {
+                filter["$and"].push({ submittedAt: { "$lte": submissionDate.to } });
+            }
+            if (submissionDate.from) {
+                filter["$and"].push({ submittedAt: { "$gte": submissionDate.from } });
+            }
+            filterQueries.push(filter);
+        }
+
+        return filterQueries;
+    }
```

Comment:
For semantic porpuses, I would separate this function into two (one which returns a single filter). I think it would make the code more readable.